berbix.go

```diff
@@ -122,11 +141,106 @@ func (c *defaultClient) UpdateTransaction(tokens *Tokens, options *UpdateTransac
+// Returns a InvalidStateErr if the upload was invalid for the current state of the transaction.
+// Returns a TransactionDoesNotExistErr if the transaction for which the image is being uploaded no longer exists.
+// Returns a PayloadTooLargeErr if the uploaded payload or underlying image is too large.
+func (c *defaultClient) UploadImage(image []byte, subject ImageSubject, format ImageFormat, tokens *Tokens) (*ImageUploadResult, error) {
+	encoded := base64.StdEncoding.EncodeToString(image)
+	req := &ImageUploadRequest{
+		Image: ImageData{
+			ImageSubject: subject,
+			Format:       format,
+			Data:         encoded,
+		},
+	}
+
+	httpResp, err := c.tokenAuthRequest(http.MethodPost, tokens, "/v0/images/upload", req)
+	if err != nil {
+		return nil, err
+	}
+	defer httpResp.Body.Close()
+	bodyDec := json.NewDecoder(httpResp.Body)
+	switch httpResp.StatusCode {
+	case http.StatusOK, http.StatusUnprocessableEntity:
+		imageResp := ImageUploadResponse{}
+		if err := bodyDec.Decode(&imageResp); err != nil {
+			return nil, fmt.Errorf("error unmarshalling response: %v", err)
+		}
+		return &ImageUploadResult{
+			IsAcceptableIDType:  httpResp.StatusCode != http.StatusUnprocessableEntity,
+			ImageUploadResponse: imageResp,
+		}, nil
+	case http.StatusConflict:
+		invalidStateRes := InvalidUploadForStateResponse{}
+		if err := bodyDec.Decode(&invalidStateRes); err != nil {
+			return nil, fmt.Errorf("got malformed response body for status code %d", httpResp.StatusCode)
+		}
+
+		return nil, InvalidStateErr{
+			InvalidUploadForStateResponse: invalidStateRes,
+		}
+	case http.StatusGone:
+		const defaultMsg = "The transaction for this upload does not exist. It may have been deleted."
+		msgErr := makeErrorMessage(bodyDec, defaultMsg)
+		return nil, TransactionDoesNotExistErr{errorMessage: msgErr}
+	case http.StatusRequestEntityTooLarge:
+		const defaultMsg = "Request body or image too large."
+		msgErr := makeErrorMessage(bodyDec, defaultMsg)
+		return nil, PayloadTooLargeErr{errorMessage: msgErr}
+	default:
+		return nil, GenericErr{
+			StatusCode: httpResp.StatusCode,
+			Message:    fmt.Sprintf("got unexpected response code %d", httpResp.StatusCode),
+		}
+	}
+}
+
+func makeErrorMessage(bodyDec *json.Decoder, defaultMsg string) errorMessage {
+	var msg string
+	genErr := GenericErrorResponse{}
+	if err := bodyDec.Decode(&genErr); err == nil {
+		msg = genErr.Message
+	} else {
+		msg = defaultMsg
+	}
+	return errorMessage{Message: msg}
+}
+
+type ImageQualityResponse struct {
+	SufficientQuality bool `json:"sufficient_quality"`
+}
+
+func (c *defaultClient) ImageQualityCheck(image []byte, subject ImageSubject, format ImageFormat) (bool, error) {
+	encoded := base64.StdEncoding.EncodeToString(image)
+	req := &ImageUploadRequest{
+		Image: ImageData{
+			ImageSubject: subject,
+			Format:       format,
+			Data:         encoded,
+		},
+	}
+
+	const uploadPath = "/v0/images/quality"
```

Comment:
does this endpoint exist yet?

Reply:
Whoops, this was a holdover from local testing on the image quality endpoint we hacked on